tools/build_langserver/langserver/analyzer.go

```diff
@@ -139,35 +170,99 @@ func (a *Analyzer) IdentFromFile(uri lsp.DocumentURI) ([]*Identifier, error) {
 		return nil, err
 	}
 
-	var idents []*Identifier
+	return stmts, nil
+}
+
+// StatementFromPos returns a Statement struct with either an Identifier or asp.Expression
+func (a *Analyzer) StatementFromPos(uri lsp.DocumentURI, position lsp.Position) (*Statement, error) {
+	// Get all the statements from the build file
+	stmts, err := a.AspStatementFromFile(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	return a.getStatementByPos(stmts, position)
+}
+
+func (a *Analyzer) getStatementByPos(stmts []*asp.Statement, position lsp.Position) (*Statement, error) {
 	for _, stmt := range stmts {
-		if stmt.Ident != nil {
-			// get the identifier type
-			var identType string
-			if stmt.Ident.Action != nil {
-				if stmt.Ident.Action.Property != nil {
-					identType = "property"
-				} else if stmt.Ident.Action.Call != nil {
-					identType = "call"
-				} else if stmt.Ident.Action.Assign != nil {
-					identType = "assign"
-				} else if stmt.Ident.Action.AugAssign != nil {
-					identType = "augAssign"
-				}
+		if withInRange(stmt.Pos, stmt.EndPos, position) {
+			// get function call, assignment, and property access
+			if stmt.Ident != nil {
+				return &Statement{
+					Ident: a.identFromStatement(stmt),
+				}, nil
 			}
+			return a.statementFromAst(reflect.ValueOf(stmt), position)
+		}
+	}
+
+	return nil, nil
+}
 
-			ident := &Identifier{
-				IdentStatement: stmt.Ident,
-				Type:           identType,
-				// -1 from asp.Statement.Pos.Line, as lsp position requires zero index
-				StartLine: stmt.Pos.Line - 1,
-				EndLine:   stmt.EndPos.Line - 1,
+// statementFromAsp recursively finds asp.IdentStatement and asp.Expression in the ast
+// and returns a valid Statement pointer if within range
+func (a *Analyzer) statementFromAst(v reflect.Value, position lsp.Position) (*Statement, error) {
+	if v.Type() == reflect.TypeOf(asp.Expression{}) {
+		expr := v.Interface().(asp.Expression)
+		if withInRange(expr.Pos, expr.EndPos, position) {
+			return &Statement{
+				Expression: &expr,
+			}, nil
+		}
+	} else if v.Type() == reflect.TypeOf([]*asp.Statement{}) && v.Len() != 0 {
+		stmts := v.Interface().([]*asp.Statement)
+		return a.getStatementByPos(stmts, position)
+	} else if v.Kind() == reflect.Ptr && !v.IsNil() {
+		return a.statementFromAst(v.Elem(), position)
+	} else if v.Kind() == reflect.Slice {
+		for i := 0; i < v.Len(); i++ {
+			stmt, err := a.statementFromAst(v.Index(i), position)
+			if err != nil {
+				return nil, err
+			}
+			if stmt != nil {
+				return stmt, nil
+			}
+		}
+	} else if v.Kind() == reflect.Struct {
+		for i := 0; i < v.NumField(); i++ {
+			stmt, err := a.statementFromAst(v.Field(i), position)
+			if err != nil {
+				return nil, err
+			}
+			if stmt != nil {
+				return stmt, nil
 			}
-			idents = append(idents, ident)
 		}
 	}
+	return nil, nil
+}
+
+func (a *Analyzer) identFromStatement(stmt *asp.Statement) *Identifier {
+	// get the identifier type
+	var identType string
```

Comment:
I'm a little dubious about this; does it really help over checking the members of `stmt.Ident` directly?

I guess I'm a little anti this kind of thing since Go doesn't have first-class enums.

Reply:
Personally I think it's a bit more readable to do that instead of going to the members, I mostly use this in the hover content part, where I check the type of `Ident` to return the respective hover content.